### PR TITLE
Add dropdown with known events to "Emit Custom Event" brick

### LINKED ISSUE #6286: Add dropdown with known events to "Emit Custom Event" brick
https://pixiebrix.slack.com/archives/C042CA15AKS/p1692296981075949

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Closes #6286 

## Reviewer Tips

- In `@/components/fields/schemaFields/widgets/SchemaCustomEventWidget.tsx`, I'm importing `import { selectExtensionKnownEventNames } from "@/pageEditor/slices/editorSelectors";`. This is throwing a lint warning for `no-restricted-paths`. Is there a different approach I should be taking?

## Demo

https://www.loom.com/share/f938771e49784ca8a301b7fb46e3ed23?sid=f5238822-6b53-415a-ba9d-608401795077

## Checklist

- [ ] Add tests
- [ ] Designate a primary reviewer @twschiller

### CODE REVIEW

**Review by twschiller — APPROVED** [review]
👍  Nice - I like the ui:widget approach

File: src/components/fields/schemaFields/fieldTypeCheckers.ts [hunk 1]
```diff
@@ -58,6 +58,10 @@ export const isButtonVariantField = (uiSchema?: UiSchema) =>
   typeof uiSchema === "object" &&
   get(uiSchema, ["ui:widget"]) === "SchemaButtonVariantWidget";
 
+export const isCustomEventField = (uiSchema?: UiSchema) =>
```
Comment (twschiller) [inline src/components/fields/schemaFields/fieldTypeCheckers.ts:61]: For `ui:widget` we should move away from defining predicate fields and just have a widget registry

**Review by twschiller — COMMENTED** [review]
Not a requirement for this PR, but we likely want to modify the event name collector to also look at Custom Event Trigger Names that are configured for starter bricks

That will enable you to pick a custom event for the trigger, and then when you add an Emit Custom Event brick, that name that's expected by the trigger will show up

https://github.com/pixiebrix/pixiebrix-extension/blob/2d187ef3be1de07cadf74b3d899ba62cf3b5702d/src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts#L72-L72

**Comment by grahamlangford** [comment]
> Not a requirement for this PR, but we likely want to modify the event name collector to also look at Custom Event Trigger Names that are configured for starter bricks
> 
> That will enable you to pick a custom event for the trigger, and then when you add an Emit Custom Event brick, that name that's expected by the trigger will show up
> 
> https://github.com/pixiebrix/pixiebrix-extension/blob/2d187ef3be1de07cadf74b3d899ba62cf3b5702d/src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts#L72-L72

I noticed that when I was testing my implementation. I'll draft a follow-on issue for it.

**Comment by Chosey98** [comment]
Great PR, I was just about to start on implementing it actually lol.

**Comment by Chosey98** [comment]
Looking forward for its release.

File: src/components/fields/schemaFields/fieldTypeCheckers.ts [hunk 2]
```diff
@@ -54,9 +54,8 @@ export const isHeadingStyleField = (fieldDefinition: Schema) =>
   fieldDefinition.type === "string" &&
   fieldDefinition.format === "heading-style";
 
-export const isButtonVariantField = (uiSchema?: UiSchema) =>
-  typeof uiSchema === "object" &&
-  get(uiSchema, ["ui:widget"]) === "SchemaButtonVariantWidget";
+export const isUiWidget = (uiSchema?: UiSchema) =>
```
Comment (twschiller) [inline src/components/fields/schemaFields/fieldTypeCheckers.ts:57]: This naming is confusing to me - I think "hasCustomWidget" or similar would clarify the usage. The uiSchema itself isn't a widget

File: src/components/fields/schemaFields/ButtonVariantSchemaField.tsx [hunk 3]
```diff
@@ -15,7 +15,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import { defaultFieldFactory } from "@/components/fields/schemaFields/SchemaFieldContext";
+import { defaultFieldFactory } from "./defaultFieldFactory";
```
Comment (twschiller) [inline src/components/fields/schemaFields/ButtonVariantSchemaField.tsx:18]: NIT: prefer absolute imports

File: src/components/fields/schemaFields/SchemaFieldContext.tsx [hunk 4]
```diff
@@ -15,45 +15,25 @@
...
-        description={description ?? schema.description}
-        as={Widget}
-      />
-    );
-  };
-
-  Field.displayName = `SchemaField(${Widget.displayName})`;
-  return Field;
-}
+export const customWidgets: CustomWidgetRegistry = {
```
Comment (twschiller) [inline src/components/fields/schemaFields/SchemaFieldContext.tsx:26]: The naming here is confusing - it's a widget registry, but the entries are fields.

My understanding is:
- Field = the label, description, widget (or composite widget)
- Widget = just the value control

File: src/components/fields/schemaFields/schemaFieldTypes.ts [hunk 5]
```diff
@@ -17,11 +17,19 @@
...
 import { type InputModeOption } from "@/components/fields/schemaFields/widgets/templateToggleWidgetTypes";
+import { type SchemaFieldComponent } from "@/components/fields/schemaFields/propTypes";
 
 export type CustomFieldToggleMode = {
   match: (fieldSchema: Schema) => boolean;
   option: InputModeOption;
 };
+
+export type CustomWidgetRegistry = {
+  ButtonVariantSchemaField: SchemaFieldComponent;
```
Comment (twschiller) [inline src/components/fields/schemaFields/schemaFieldTypes.ts:28]: See comment on naming fields vs. widget here: https://github.com/pixiebrix/pixiebrix-extension/pull/6288/files#r1300729929

**Review by twschiller — COMMENTED** [review]
@grahamlangford overall it looks good, but let's sort out the widget vs. field terminology

File: src/components/fields/schemaFields/ButtonVariantSchemaField.tsx (hunk 3 above)
Comment (grahamlangford) [reply src/components/fields/schemaFields/ButtonVariantSchemaField.tsx:18]: Good to know. Depends on the project, but I guess I've noticed that. I know there's a linting rule for that. I'll add it to the list for the next time I'm on a DX week.

https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/prefer-default-export.md

**Pull request merged**